File: app/src/main/java/com/webviewblock/presentation/settings/adapter/HistoryAdapter.java
package com.webviewblock.presentation.settings.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.webviewblock.databinding.LayoutEmptyHistoryBinding;
import com.webviewblock.databinding.RowHistoryBinding;
import com.webviewblock.domain.History;

import java.util.ArrayList;
import java.util.List;

public class HistoryAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final int VIEW_TYPE_ITEM = 0;
    private final int VIEW_TYPE_EMPTY_STATE = 1;
    private List<History> items = null;
    private ClickListenerHistory listener;

    @Override
    public int getItemViewType(int position) {
        if (items.isEmpty())
            return VIEW_TYPE_EMPTY_STATE;
        return VIEW_TYPE_ITEM;
    }

    public interface ClickListenerHistory {
        void clickItem(String url);
    }

    public void setup(List<History> newItems, ClickListenerHistory listener) {
        if(items != null)
            items.clear();
        else items = new ArrayList<>();
        this.listener = listener;
        if (!newItems.isEmpty()) {
            items.addAll(newItems);
        }
        notifyDataSetChanged();
    }

    public void setItems(List<History> newItems) {
        if(items != null)
            items.clear();
        else items = new ArrayList<>();
        if (!newItems.isEmpty()) {
            items.addAll(newItems);
        }
        notifyDataSetChanged();
    }

    private void clear() {
        if(items != null)
            items.clear();
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case VIEW_TYPE_ITEM:
                RowHistoryBinding binding = RowHistoryBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
                return new MessageViewHolder(binding);
            default:
                LayoutEmptyHistoryBinding bindingHist = LayoutEmptyHistoryBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
                return new EmptyViewHolder(bindingHist);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof MessageViewHolder) {
            History item = items.get(position);
            //url
            MessageViewHolder viewHolder = (MessageViewHolder) holder;
            viewHolder.binding.rowHistoryTextviewUrl.setText(item.getUrl());
            //date
            viewHolder.binding.rowHistoryTextviewDate.setText(item.getTime());
            //click
                viewHolder.binding.getRoot().setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        listener.clickItem(item.getUrl());
                    }
                });
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    private class EmptyViewHolder extends RecyclerView.ViewHolder {

        public EmptyViewHolder(LayoutEmptyHistoryBinding binding) {
            super(binding.getRoot());
        }
    }

    private class MessageViewHolder extends RecyclerView.ViewHolder {

        RowHistoryBinding binding;

        public MessageViewHolder(RowHistoryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/webviewblock/presentation/settings/adapter/HistoryAdapter.kt
package com.webviewblock.presentation.settings.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.webviewblock.databinding.LayoutEmptyHistoryBinding
import com.webviewblock.databinding.RowHistoryBinding
import com.webviewblock.domain.History

class HistoryAdapter(
    private val clickListener: (item: History) -> Unit
) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    var items: MutableList<History> = emptyList<History>().toMutableList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    fun setup(newItems: List<History>) {
        _showError = false
        items.clear()
        if (newItems.isNotEmpty()) {
            items.addAll(newItems)
        }
        notifyDataSetChanged()
    }

    fun clear() {
        items.clear()
        _firstLoading = true
        notifyDataSetChanged()
    }

    internal class EmptyViewHolder(val binding: LayoutEmptyHistoryBinding) :
        RecyclerView.ViewHolder(binding.root)

    internal class MessageViewHolder(val binding: RowHistoryBinding) :
        RecyclerView.ViewHolder(binding.root)

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is MessageViewHolder) {
            val item = items[position]

            if (holder is MessageViewHolder) {
                //url
                val viewHolder = holder as MessageViewHolder
                viewHolder.binding.rowHistoryTextviewUrl.text = item.url
                //date
                viewHolder.binding.rowHistoryTextviewDate.text = item.time
                //click
                viewHolder.binding.root.setOnClickListener {
                    clickListener.invoke(item)
                }
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (items.isEmpty())
            return VIEW_TYPE_EMPTY_STATE
        return VIEW_TYPE_ITEM
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            VIEW_TYPE_ITEM -> {
                val binding = RowHistoryBinding
                    .inflate(LayoutInflater.from(parent.context), parent, false)
                return MessageViewHolder(binding)
            }
            else -> {
                val binding = LayoutEmptyHistoryBinding
                    .inflate(LayoutInflater.from(parent.context), parent, false)
                return EmptyViewHolder(binding)
            }
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    private var _firstLoading = false
    private var _showError = false

    companion object {
        private const val VIEW_TYPE_ITEM = 0
        private const val VIEW_TYPE_EMPTY_STATE = 1
    }
}
